ckanext/qdes/logic/action/get.py

```diff
@@ -0,0 +1,341 @@
+            rows.append({
+                'Dataset name': package.get('title', package.get('name', '')),
+                'Link to dataset (URI)': url_for('dataset.read', id=package.get('id'), _external=True),
+                'Resource name': '',
+                'Link to resource': '',
+                'Dataset creator': package.get('contact_creator', ''),
+                'Point of contact - name': _qdes_extract_point_of_contact(package.get('contact_point', None), 'Name'),
+                'Point of contact - email': _qdes_extract_point_of_contact(package.get('contact_point', None), 'Email'),
+                'List of recommended fields without values': ', '.join(missing_values),
+                'Organisation name': package.get('organization').get('title', ''),
+            })
+
+            # Check dataset resource metadata fields.
+            for resource in package.get('resources', []):
+                missing_values = _check_recommended_field_value(resource, dataset_resource_recommended_fields)
+                rows.append({
+                    'Dataset name': package.get('title', package.get('name', '')),
+                    'Link to dataset (URI)': url_for('dataset.read', id=package.get('id'), _external=True),
+                    'Resource name': resource.get('name', ''),
+                    'Link to resource': url_for('resource.read',
+                                                resource_id=resource.get('id'),
+                                                id=package.get('id'),
+                                                package_type=package.get('type'),
+                                                _external=True
+                                                ),
+                    'Dataset creator': package.get('contact_creator', ''),
+                    'Point of contact - name':
+                        _qdes_extract_point_of_contact(package.get('contact_point', None), 'Name'),
+                    'Point of contact - email':
+                        _qdes_extract_point_of_contact(package.get('contact_point', None), 'Email'),
+                    'List of recommended fields without values': ', '.join(missing_values),
+                    'Organisation name': package.get('organization').get('title', ''),
+                })
+
+    return rows
+
+
+def _get_uri_validated_fields(scheme, field_group):
+    uri_validated_fields = []
+    for field in scheme.get(field_group, []):
+        validator = field.get('validators', '').split()
+        if 'qdes_uri_validator' in validator:
+            uri_validated_fields.append(field)
+
+    return uri_validated_fields
+
+
+def qdes_datasets_with_invalid_urls(context, config):
+    u"""
+    List of all datasets with broken links to resources.
+    """
+    # Get list of invalid uris.
+    uris = Session.query(InvalidUri).all()
+    invalid_uris = [uri.as_dict() for uri in uris]
```

Comment:
Hi @awang - converting each of the `InvalidUri` objects may not be necessary here as we can just iterate through the objects in the `for invalid_uri in invalid_uris:` block below and use the object's properties to populate the `entities` dict... I will add an example below.

Reply:
ok noted, will implement for others.